firedrake/preconditioners/patch.py

```diff
@@ -568,11 +571,41 @@ def select_entity(p, dm=None, exclude=None):
 class PlaneSmoother(object):
     @staticmethod
     def coords(dm, p):
-        coordsSection = dm.getCoordinateSection()
-        coordsDM = dm.getCoordinateDM()
-        dim = coordsDM.getDimension()
-        coordsVec = dm.getCoordinatesLocal()
-        return dm.getVecClosure(coordsSection, coordsVec, p).reshape(-1, dim).mean(axis=0)
+        mesh = dm.getAttr("__firedrake_mesh__")
+
+        ele = mesh.coordinates.function_space().ufl_element()
+        assert ele.family() in ["Lagrange", "Discontinuous Lagrange"]
+        if ele.family() == "Discontinuous Lagrange":
+            # Ah, we got a periodic mesh. We need to interpolate to CGk
+            # with access descriptor MAX to define a consistent opinion
+            # about where the vertices are.
+
+            coords = dm.getAttr("__firedrake_interpolated_coords__")
+            if coords is None:
+                CGkele = ele.reconstruct(family="Lagrange")
+                # Can't do
+                # CGk = FunctionSpace(mesh, CGkele)
+                # because equality fails between a mesh and a weakref.proxy.
+                CGk = FunctionSpace(mesh.coordinates.function_space().mesh(), CGkele)
+                coords = interpolate(mesh.coordinates, CGk, access=op2.MAX)
+                dm.setAttr("__firedrake_interpolated_coords__", coords)
+        else:
+            coords = mesh.coordinates
+
+        coordsV = coords.function_space()
+        data = coords.dat.data_ro
```

Comment:
I think this will fail in parallel because if `p` is a non-ghost cell, the closure may contain points that are `ghost` which will then break.

Reply:
Thanks for pointing this out. I've switched it to use `data_ro_with_halos`.